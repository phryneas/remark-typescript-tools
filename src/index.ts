export {
  attacher as linkDocblocks,
  Settings as LinkDocblocksSettings,
} from './linkDocblocks/plugin';

export {
  attacher as transpileCodeblocks,
  Settings as TranspileCodeblocksSettings,
  defaultAssembleReplacementNodes,
} from './transpileCodeblocks/plugin';

export {
  postProcessTranspiledJs as defaultPostProcessTranspiledJs,
  postProcessTs as defaultPostProcessTs,
} from './transpileCodeblocks/postProcessing';
