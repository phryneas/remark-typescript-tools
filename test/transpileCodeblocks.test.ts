import unified from 'unified';
import remarkParser from 'remark-parse';
// @ts-ignore
import mdxPlugin from 'remark-mdx';
import remarkStringify from 'remark-stringify';
import { transpileCodeblocks, TranspileCodeblocksSettings } from '../src';
import { resolve } from 'path';

expect.addSnapshotSerializer({
  test(value) {
    return value && value.type;
  },
  print(value, serialize) {
    if (value.type === 'code') {
      return (
        '```' +
        value.lang +
        (value.meta ? ' ' + value.meta : '') +
        '\n' +
        value.value +
        '\n```'
      );
    }
    return value.children
      ? value.children.map(serialize).join('\n')
      : value.value;
  },
});

const defaultSettings: TranspileCodeblocksSettings = {
  compilerSettings: {
    tsconfig: resolve(__dirname, '..', 'tsconfig.tests.json'),
    externalResolutions: {
      'remark-typescript-tools': {
        resolvedPath: resolve(__dirname, '..', 'src'),
        packageId: {
          name: 'remark-typescript-tools',
          subModuleName: '',
          version: '1.0',
        },
      },
    },
  },
};

function getParser(settings = defaultSettings) {
  return unified()
    .use(remarkParser)
    .use(mdxPlugin)
    .use(transpileCodeblocks, settings)
    .use(remarkStringify);
}

function transform(md: string, parser = getParser()) {
  return parser.run(parser.parse(md), {
    path: __dirname + '/test.mdx',
  });
}

test('transpiles codeblocks', async () => {
  const md = `
\`\`\`ts
function testFn(arg1: string) {
    return arg1;
}
\`\`\`
`;

  expect(await transform(md)).toMatchSnapshot();
});

test('inserts imports for Tabs & TabItem', async () => {
  const md = `
# some-pararaph
`;

  expect(await transform(md)).toMatchInlineSnapshot(`
    import TabItem from '@theme/TabItem'
    import Tabs from '@theme/Tabs'
    some-pararaph
  `);
});

test('skips imports that are already present', async () => {
  const md = `
# some-pararaph

import Tabs from '@theme/Tabs'
`;

  expect(await transform(md)).toMatchInlineSnapshot(`
    import TabItem from '@theme/TabItem'
    some-pararaph
    import Tabs from '@theme/Tabs'

  `);
});

test('throws an error on compilation error', async () => {
  const md = `
\`\`\`ts
let x: string = 5
\`\`\`
`;

  await expect(
    transform(md).catch((e) => {
      throw e.toString();
    })
  ).rejects.toContain(
    `remark-typescript-tools/test/test.mdx:2:1-4:4: Type '5' is not assignable to type 'string'.`
  );
});

test('no-transpile does not throw an error and does add tabs', async () => {
  const md = `
\`\`\`ts no-transpile
let x: string = 5
\`\`\`
`;

  await expect(transform(md)).resolves.toMatchInlineSnapshot(`
    import TabItem from '@theme/TabItem'
    import Tabs from '@theme/Tabs'
    \`\`\`ts
    let x: string = 5
    \`\`\`
  `);
});

test('transpiles codeblocks with multiple file definitions', async () => {
  const md = `
\`\`\`ts
// file: file1.ts
export function testFn(arg1: string) {
    return arg1;
}
// file: file2.ts
import { testFn } from './file1'

console.log(testFn("foo"))
\`\`\`
`;

  expect(await transform(md)).toMatchSnapshot();
});

test('finds error spanning over multiple files', async () => {
  const md = `
\`\`\`ts
// file: file1.ts
export function testFn(arg1: string) {
    return arg1;
}
// file: file2.ts
import { testFn } from './file1'

console.log(testFn(5))
\`\`\`
`;

  await expect(
    transform(md).catch((e) => {
      throw e.toString();
    })
  ).rejects
    .toContain(`/remark-typescript-tools/test/test.mdx/codeBlock_1/file2.ts
Argument of type '5' is not assignable to parameter of type 'string'.`);
});

test('takes "noEmit" files into account for compiling, but does not output them', async () => {
  const md = `
\`\`\`ts
// file: file1.ts noEmit
export function testFn(arg1: string) {
    return arg1;
}
// file: file2.ts
import { testFn } from './file1'

console.log(testFn("foo"))
\`\`\`
`;

  expect(await transform(md)).toMatchSnapshot(
    'file1.ts should be missing from this snapshot'
  );
});

test('reports errors in "noEmit" files', async () => {
  const md = `
\`\`\`ts
// file: file1.ts noEmit
export function testFn(arg1: string) {
    let x: number = arg1;
    return x;
}
// file: file2.ts
import { testFn } from './file1'

console.log(testFn("foo"))
\`\`\`
`;

  await expect(
    transform(md).catch((e) => {
      throw e.toString();
    })
  ).rejects
    .toContain(`remark-typescript-tools/test/test.mdx/codeBlock_1/file1.ts
Type 'string' is not assignable to type 'number'.`);
});

test('supports hyphens, square brackets & periods in filenames', async () => {
  const md = `
\`\`\`ts
// file: file-one.stuff.ts noEmit
export function testFn(arg1: string) {
    return arg1;
}
// file: [file2].ts
import { testFn } from './file-one.stuff'

console.log(testFn("foo"))
\`\`\`
`;

  expect(await transform(md)).toMatchSnapshot();
});
export const someNumber = 5 as const;

describe('imports defined via compilerOptions.paths', () => {
  test('import', async () => {
    const md = `
\`\`\`ts
// file: file1.ts
import { someNumber } from "@transpileCodeblocksTest"
const n: number = someNumber;
// file: file2.ts
import { someNumber } from "@test/transpileCodeblocks.test"
const n: number = someNumber;
\`\`\`
`;
    expect(await transform(md)).toMatchSnapshot();
  });
  test('import with errors', async () => {
    const md = `
\`\`\`ts
import { someNumber } from "@test/transpileCodeblocks.test"
const n: string = someNumber;
\`\`\`
`;
    await expect(
      transform(md).catch((e) => {
        throw e.toString();
      })
    ).rejects
      .toContain(`remark-typescript-tools/test/test.mdx/codeBlock_1/index.ts
Type '5' is not assignable to type 'string'.`);
  });
});

test('transpiles jsx', async () => {
  const md = `
\`\`\`ts
// file: file2.tsx
import React from 'react';

console.log(<div>asd</div>)
\`\`\`
`;

  expect(await transform(md)).toMatchSnapshot();
});

test('transpiles multiple jsx files', async () => {
  const md = `
\`\`\`ts
// file: button.tsx
import React from 'react';

export const Button = (props: React.ButtonHTMLAttributes<HTMLButtonElement>) => <button {...props} />;

// file: file2.tsx
import React from 'react';
import { Button } from "./button";

console.log(<Button>asd</Button>);

\`\`\`
`;
  expect(await transform(md)).toMatchSnapshot();
});

test('transforms virtual filepath', async () => {
  const md = `
  \`\`\`ts
  // file: file1.ts
  export function testFn(arg1: string) {
      return arg1;
  }
  // file: file2.ts
  import { testFn } from './file1'
  
  console.log(testFn(5))
  \`\`\`
  `;

  const settings: TranspileCodeblocksSettings = {
    ...defaultSettings,
    compilerSettings: {
      ...defaultSettings.compilerSettings,
      transformVirtualFilepath: (path) =>
        path.replace('/test/', '/replaced/path/'),
    },
  };

  const parser = getParser(settings);

  await expect(
    transform(md, parser).catch((e) => {
      throw e.toString();
    })
  ).rejects
    .toContain(`/remark-typescript-tools/replaced/path/test.mdx/codeBlock_1/file2.ts
Argument of type '5' is not assignable to parameter of type 'string'.`);
});

test('supports tsx snippets', async () => {
  const md = `
\`\`\`tsx title="App.tsx" showLineNumbers
// file: App.tsx
import React from 'react';
export function App() {
  const [counter, setCounter] = React.useState<number>(0);
  return (
    <div>
      <button onClick={() => setCounter((prev) => prev + 1)}>
        Increment counter ({counter})
      </button>
    </div>
  )
}
`;

  expect(await transform(md)).toMatchSnapshot();
});
