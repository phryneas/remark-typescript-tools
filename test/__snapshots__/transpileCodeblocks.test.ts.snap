// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imports defined via compilerOptions.paths import 1`] = `
import TabItem from '@theme/TabItem'
import Tabs from '@theme/Tabs'

    <Tabs
      groupId="language"
      defaultValue="ts"
      values={[
        { label: 'TypeScript', value: 'ts', },
        { label: 'JavaScript', value: 'js', },
      ]}
    >        
        <TabItem value="ts">
\`\`\`ts
// file: file1.ts
import { someNumber } from '@transpileCodeblocksTest';
const n: number = someNumber;


// file: file2.ts
import { someNumber } from '@test/transpileCodeblocks.test';
const n: number = someNumber;
\`\`\`

        </TabItem>
        <TabItem value="js">
\`\`\`js
// file: file1.js
import { someNumber } from '@transpileCodeblocksTest';
const n = someNumber;


// file: file2.js
import { someNumber } from '@test/transpileCodeblocks.test';
const n = someNumber;
\`\`\`

        </TabItem>
    </Tabs>
`;

exports[`supports hyphens, square brackets & periods in filenames 1`] = `
import TabItem from '@theme/TabItem'
import Tabs from '@theme/Tabs'

    <Tabs
      groupId="language"
      defaultValue="ts"
      values={[
        { label: 'TypeScript', value: 'ts', },
        { label: 'JavaScript', value: 'js', },
      ]}
    >        
        <TabItem value="ts">
\`\`\`ts
import { testFn } from './file-one.stuff';

console.log(testFn('foo'));
\`\`\`

        </TabItem>
        <TabItem value="js">
\`\`\`js
import { testFn } from './file-one.stuff';

console.log(testFn('foo'));
\`\`\`

        </TabItem>
    </Tabs>
`;

exports[`supports tsx snippets 1`] = `
import TabItem from '@theme/TabItem'
import Tabs from '@theme/Tabs'

    <Tabs
      groupId="language"
      defaultValue="ts"
      values={[
        { label: 'TypeScript', value: 'ts', },
        { label: 'JavaScript', value: 'js', },
      ]}
    >        
        <TabItem value="ts">
\`\`\`tsx title="App.tsx" showLineNumbers
import React from 'react';
export function App() {
  const [counter, setCounter] = React.useState<number>(0);
  return (
    <div>
      <button onClick={() => setCounter((prev) => prev + 1)}>
        Increment counter ({counter})
      </button>
    </div>
  );
}
\`\`\`

        </TabItem>
        <TabItem value="js">
\`\`\`js title="App.jsx" showLineNumbers
import React from 'react';
export function App() {
  const [counter, setCounter] = React.useState(0);
  return (
    <div>
      <button onClick={() => setCounter((prev) => prev + 1)}>
        Increment counter ({counter})
      </button>
    </div>
  );
}
\`\`\`

        </TabItem>
    </Tabs>
`;

exports[`takes "noEmit" files into account for compiling, but does not output them: file1.ts should be missing from this snapshot 1`] = `
import TabItem from '@theme/TabItem'
import Tabs from '@theme/Tabs'

    <Tabs
      groupId="language"
      defaultValue="ts"
      values={[
        { label: 'TypeScript', value: 'ts', },
        { label: 'JavaScript', value: 'js', },
      ]}
    >        
        <TabItem value="ts">
\`\`\`ts
import { testFn } from './file1';

console.log(testFn('foo'));
\`\`\`

        </TabItem>
        <TabItem value="js">
\`\`\`js
import { testFn } from './file1';

console.log(testFn('foo'));
\`\`\`

        </TabItem>
    </Tabs>
`;

exports[`transpiles codeblocks 1`] = `
import TabItem from '@theme/TabItem'
import Tabs from '@theme/Tabs'

    <Tabs
      groupId="language"
      defaultValue="ts"
      values={[
        { label: 'TypeScript', value: 'ts', },
        { label: 'JavaScript', value: 'js', },
      ]}
    >        
        <TabItem value="ts">
\`\`\`ts
function testFn(arg1: string) {
  return arg1;
}
\`\`\`

        </TabItem>
        <TabItem value="js">
\`\`\`js
function testFn(arg1) {
  return arg1;
}
\`\`\`

        </TabItem>
    </Tabs>
`;

exports[`transpiles codeblocks with multiple file definitions 1`] = `
import TabItem from '@theme/TabItem'
import Tabs from '@theme/Tabs'

    <Tabs
      groupId="language"
      defaultValue="ts"
      values={[
        { label: 'TypeScript', value: 'ts', },
        { label: 'JavaScript', value: 'js', },
      ]}
    >        
        <TabItem value="ts">
\`\`\`ts
// file: file1.ts
export function testFn(arg1: string) {
  return arg1;
}


// file: file2.ts
import { testFn } from './file1';

console.log(testFn('foo'));
\`\`\`

        </TabItem>
        <TabItem value="js">
\`\`\`js
// file: file1.js
export function testFn(arg1) {
  return arg1;
}


// file: file2.js
import { testFn } from './file1';

console.log(testFn('foo'));
\`\`\`

        </TabItem>
    </Tabs>
`;

exports[`transpiles jsx 1`] = `
import TabItem from '@theme/TabItem'
import Tabs from '@theme/Tabs'

    <Tabs
      groupId="language"
      defaultValue="ts"
      values={[
        { label: 'TypeScript', value: 'ts', },
        { label: 'JavaScript', value: 'js', },
      ]}
    >        
        <TabItem value="ts">
\`\`\`ts
import React from 'react';

console.log(<div>asd</div>);
\`\`\`

        </TabItem>
        <TabItem value="js">
\`\`\`js
import React from 'react';

console.log(<div>asd</div>);
\`\`\`

        </TabItem>
    </Tabs>
`;

exports[`transpiles multiple jsx files 1`] = `
import TabItem from '@theme/TabItem'
import Tabs from '@theme/Tabs'

    <Tabs
      groupId="language"
      defaultValue="ts"
      values={[
        { label: 'TypeScript', value: 'ts', },
        { label: 'JavaScript', value: 'js', },
      ]}
    >        
        <TabItem value="ts">
\`\`\`ts
// file: button.tsx
import React from 'react';

export const Button = (
  props: React.ButtonHTMLAttributes<HTMLButtonElement>
) => <button {...props} />;


// file: file2.tsx
import React from 'react';
import { Button } from './button';

console.log(<Button>asd</Button>);
\`\`\`

        </TabItem>
        <TabItem value="js">
\`\`\`js
// file: button.jsx
import React from 'react';

export const Button = (props) => <button {...props} />;


// file: file2.jsx
import React from 'react';
import { Button } from './button';

console.log(<Button>asd</Button>);
\`\`\`

        </TabItem>
    </Tabs>
`;
